crypto/x509/v3_rolespec.c

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/asn1t.h>
+#include <openssl/x509v3.h>
+#include "ext_dat.h"
+
+ASN1_SEQUENCE(OSSL_ROLE_SPEC_CERT_ID) = {
+    ASN1_EXP(OSSL_ROLE_SPEC_CERT_ID, roleName, GENERAL_NAME, 0),
+    ASN1_EXP(OSSL_ROLE_SPEC_CERT_ID, roleCertIssuer, GENERAL_NAME, 1),
+    ASN1_IMP_OPT(OSSL_ROLE_SPEC_CERT_ID, roleCertSerialNumber, ASN1_INTEGER, 2),
+    ASN1_IMP_SEQUENCE_OF_OPT(OSSL_ROLE_SPEC_CERT_ID, roleCertLocator, GENERAL_NAME, 3),
+} ASN1_SEQUENCE_END(OSSL_ROLE_SPEC_CERT_ID)
+
+IMPLEMENT_ASN1_FUNCTIONS(OSSL_ROLE_SPEC_CERT_ID)
+
+ASN1_ITEM_TEMPLATE(OSSL_ROLE_SPEC_CERT_ID_SYNTAX) =
+    ASN1_EX_TEMPLATE_TYPE(ASN1_TFLG_SEQUENCE_OF,
+                          0, OSSL_ROLE_SPEC_CERT_ID_SYNTAX, OSSL_ROLE_SPEC_CERT_ID)
+ASN1_ITEM_TEMPLATE_END(OSSL_ROLE_SPEC_CERT_ID_SYNTAX)
+
+IMPLEMENT_ASN1_FUNCTIONS(OSSL_ROLE_SPEC_CERT_ID_SYNTAX)
+
+static int integer_print_bio(BIO *bio, const ASN1_INTEGER *num)
+{
+    BIGNUM *num_bn;
+    int result = 0;
+    char *hex;
+
+    num_bn = ASN1_INTEGER_to_BN(num, NULL);
+    if (num_bn == NULL)
+        return -1;
+    if ((hex = BN_bn2hex(num_bn))) {
+        result = BIO_write(bio, "0x", 2) > 0;
+        result = result && BIO_write(bio, hex, strlen(hex)) > 0;
+        OPENSSL_free(hex);
+    }
+    BN_free(num_bn);
+    return result;
+}
```

Comment:
I see this is being copied from TS_ASN1_INTEGER_print_bio(). It should be factorized out into a common internal routine.

However the certificate serial numbers are printed in https://github.com/openssl/openssl/blob/master/crypto/x509/t_x509.c#L88-L121

Would it make sense to use the same routine?

I am not sure which one is better though.

Reply:
I think it does make sense to use the same routine. The serial number printing function prints the number in decimal and hex and alerts users about negative values. It seems all around better to me, and lots of extensions and attribute certificates use serial numbers other than in the `serialNumber` field itself, so it makes sense to have a consistent output for these.

Reply:
I defined `ossl_serial_number_print` and made both the `roleSpecCertIdentifier` extension and `X509_print_ex` use it.